api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+  map<string, InputArtifactSpec> artifacts = 2;
+}
+
+// The spec of task outputs.
+message TaskOutputsSpec {
+  // The specification of a task output artifact.
+  message OutputArtifactSpec {
+    // The type of the artifact.
+    ArtifactTypeSchema artifact_type = 1;
+
+    // The properties of the artifact, which are determined either at
+    // compile-time, or at pipeline submission time through runtime parameters
+    map<string, ValueOrRuntimeParameter> properties = 2;
+
+    // The custom properties of the artifact, which are determined either at
+    // compile-time, or at pipeline submission time through runtime parameters
+    map<string, ValueOrRuntimeParameter> custom_properties = 3;
+  }
+
+  // Specification for output parameters produced by the task.
+  message OutputParameterSpec {
+    // Required field. The type of the output parameter.
+    PrimitiveType.PrimitiveTypeEnum type = 1;
+  }
+
+  // A map of output parameters which are small values, stored by the system and
+  // can be queriable. The output key is used
+  // by [TaskInputsSpec.InputParameterSpec][] of the downstream task to specify
+  // the data dependency. The same key will also be used by
+  // [ExecutorInput.Inputs][] to reference the output parameter.
+  map<string, OutputParameterSpec> parameters = 1;
+  // A map of output artifacts. Keyed by output key. The output key is used
+  // by [TaskInputsSpec.InputArtifactSpec][] of the downstream task to specify
+  // the data dependency. The same key will also be used by
+  // [ExecutorInput.Inputs][] to reference the output artifact.
+  map<string, OutputArtifactSpec> artifacts = 2;
+}
+
+// Represent primitive types. The wrapper is needed to give a namespace of
+// enum value so we don't need add `PRIMITIVE_TYPE_` prefix of each enum value.
+message PrimitiveType {
+  // The primitive types.
+  enum PrimitiveTypeEnum {
+    PRIMITIVE_TYPE_UNSPECIFIED = 0;
+    INT = 1;
+    DOUBLE = 2;
+    STRING = 3;
+  }
+}
+
+// The spec of a pipeline task.
+message PipelineTaskSpec {
+  // Basic info of a pipeline task.
+  PipelineTaskInfo task_info = 1;
+
+  // Specification for task inputs which contains parameters and artifacts.
+  TaskInputsSpec inputs = 2;
+
+  // Specification for task outputs.
+  TaskOutputsSpec outputs = 3;
+
+  // Label for the executor of the task.
+  // The specification will be specified in the deployment config.
+  // For example:
+  // ```
+  // tasks:
+  // - task_info:
+  //     name: trainer
+  //   executor_label: trainer
+  // deployment_config:
+  //   @type: cloud.ml.pipelines.v1alpha3.proto.PipelineDeploymentConfig
+  //   executors:
+  //     trainer:
+  //       container:
+  //         image: gcr.io/tfx:latest
+  //         args: []
+  // ```
+  string executor_label = 4;
+
+  // A list of names of upstream tasks that do not provide input
+  // artifacts for this task, but nonetheless whose completion this task depends
+  // on.
+  repeated string dependent_tasks = 5;

```

Comment:
Nit: I think "dependent_tasks" means downstream tasks. You might want to use "dependencies" or "upstream_tasks" or "additional_dependencies" (which implies that for normal data dependencies this field can be left empty).

Reply:
Hmm, i was considering this as the same as dependencies. But since we also have data dependency, it will be confusing to call it "dependencies". I need a second opinion on this naming. @james-jwu @neuromage